File: Leetcode/typescript/214-shortest-palindrome.ts
function shortestPalindrome(s: string): string {
  function isPalindrome(s: string, l: number, r: number): boolean {
    while (l <= r) {
      if (s[l] !== s[r]) return false;
      l++;
      r--;
    }
    return true;
  }
  const n = s.length;
  let r: number;
  for (r = n; r >= 0; r--) {
    if (isPalindrome(s, 0, r)) break;
  }
  const suffix = s.slice(r + 1, n);

  return suffix.split("").reverse().join("") + s;
}

File: Leetcode/python/214-shortest-palindrome.py
class Solution:
    def shortestPalindrome(self, s: str) -> str:
        def isPalindrome(s, l, r):
            while l <= r:
                if s[l] != s[r]:
                    return False
                l, r = l + 1, r - 1
            return True
        for r in reversed(range(len(s))):
            if isPalindrome(s, 0, r):
                suffix = s[r + 1:]
                return suffix[::-1] + s
        return ""